ensembleperturbation/uncertainty_quantification/surrogate.py

```diff
@@ -606,3 +606,154 @@ def compute_surrogate_percentiles(
     out = out.reshape(q.shape + shape)
 
     return out
+
+
+def probability_field_from_samples(
+    samples: xarray.Dataset,
+    levels: List[float],
+    surrogate_model: numpoly.ndpoly,
+    distribution: chaospy.Distribution,
+    minimum_allowable_value: float = None,
+    convert_from_log_scale: Union[bool, float] = False,
+    convert_from_depths: Union[bool, float] = False,
+) -> xarray.DataArray:
+
+    LOGGER.info(f'calculating {len(levels)} probability field(s): {levels}')
+
+    surrogate_prob_field = compute_surrogate_probability_field(
+        poly=surrogate_model,
+        levels=levels,
+        dist=distribution,
+        minimum_allowable_value=minimum_allowable_value,
+        convert_from_log_scale=convert_from_log_scale,
+        convert_from_depths=convert_from_depths,
+        depths=samples['depth'],
+    )
+
+    surrogate_prob_field = xarray.DataArray(
+        surrogate_prob_field,
+        coords={
+            'level': levels,
+            **{
+                coord: values
+                for coord, values in samples.coords.items()
+                if coord not in ['run', 'type']
+            },
+        },
+        dims=('level', *(dim for dim in samples.dims if dim not in ['run', 'type'])),
+    )
+
+    return surrogate_prob_field
+
+
+def probability_field_from_surrogate(
+    levels: List[float],
+    surrogate_model: numpoly.ndpoly,
+    distribution: chaospy.Distribution,
+    training_set: xarray.Dataset,
+    minimum_allowable_value: float = None,
+    convert_from_log_scale: Union[bool, float] = False,
+    convert_from_depths: Union[bool, float] = False,
+    element_table: xarray.DataArray = None,
+    filename: PathLike = None,
+) -> xarray.Dataset:
+
+    if filename is not None and not isinstance(filename, Path):
+        filename = Path(filename)
+
+    if filename is None or not filename.exists():
+        surrogate_prob_field = probability_field_from_samples(
+            samples=training_set,
+            levels=levels,
+            surrogate_model=surrogate_model,
+            distribution=distribution,
+            minimum_allowable_value=minimum_allowable_value,
+            convert_from_log_scale=convert_from_log_scale,
+            convert_from_depths=convert_from_depths,
+        )
+
+        # before evaluating prob. field for model set null water elevation to the ground elevation
+        if minimum_allowable_value is not None:
+            too_small = (training_set + training_set['depth']).values < minimum_allowable_value
+            training_set.values[too_small] = numpy.nan
+        training_set = numpy.fmax(training_set, -training_set['depth'])
```

Comment:
@WPringle could the difference in the model vs surrogate be how/when we adjust?
![image](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/assets/77082694/985ea635-7fbd-490c-b44e-ba425678ae96)

Reply:
That's a good point. What was the min_depth set and what are the depths in the SCHISM model on the overland points that get wet from NWM?

Reply:
I think the mindepth is `0.01`, I need to check the depths on the overland... I'll get back to you on that. In the meantime, can you please also take a look at the newly added code as a whole and see if you spot any obvious/conceptual mistakes or not?

Reply:
@WPringle minimum allowable depth is set to `0.8 * h0` which is `0.8 * 0.01`. The depths that are set as being *too small* are:
```python
too_small = (maxelev+ depths).values < min_depth
```
These nodes are set to `nan` in the [latest version](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/pull/127/commits/5364e2c878ae3144d279fd0ec1183e024864dd0c) of the code; in an earlier version, these values would then be set to be equal to `-depth` which resulted in the earlier plots [above](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/pull/127#issuecomment-1861783391)
Looking at the `max` and `min` of the depths and maxelev of the `too_small` region show:
```
depths[too_small.any(axis=0)].min(), depths[too_small.any(axis=0)].max(), depths[too_small.any(axis=0)].mean()
# (-44.28715, -0.01106, -5.5330863)

maxelev[too_small].min(), maxelev[too_small].max(), maxelev[too_small].mean()
# (0.010003523, 44.07152, 5.2620897)
```